File: android/src/main/java/com/sourcepoint/reactnativecmp/consents/RNSPGlobalCMPConsent.kt
package com.sourcepoint.reactnativecmp.consents

import com.facebook.react.bridge.Arguments.createArray
import com.facebook.react.bridge.Arguments.createMap
import com.facebook.react.bridge.ReadableMap
import com.sourcepoint.cmplibrary.model.exposed.GlobalCmpConsent

data class RNSPGlobalCMPConsent(
  val uuid: String?,
  val applies: Boolean,
  val createdDate: String?,
  val expirationDate: String?,
  val vendors: List<RNSPConsentable>,
  val categories: List<RNSPConsentable>
): RNMappable {
  constructor(globalCMP: GlobalCmpConsent) : this(
    uuid = globalCMP.uuid,
    applies = globalCMP.applies,
    createdDate = globalCMP.dateCreated,
    expirationDate = null,
    vendors = globalCMP.vendors?.map { RNSPConsentable(it) } ?: emptyList(),
    categories = globalCMP.categories?.map { RNSPConsentable(it) } ?: emptyList()
  )

  override fun toRN(): ReadableMap = createMap().apply {
    putString("uuid", uuid)
    putBoolean("applies", applies)
    putString("createdDate", createdDate)
    putString("expirationDate", expirationDate)
    putArray("vendors", createArray().apply { vendors.forEach { pushMap(it.toRN()) } })
    putArray("categories", createArray().apply { categories.forEach { pushMap(it.toRN()) } })
  }
}

File: ios/RNSPUserData/RNSPGlobalCMPConsent.swift
//
//  RNSPGlobalCMPConsent.swift
//  sourcepoint-react-native-cmp
//
//  Created by Andre Herculano on 22/5/24.
//

import Foundation
import ConsentViewController

struct RNSPGlobalCMPConsent: Encodable {
    let uuid: String?
    let applies: Bool
    let vendors, categories: [RNSPConsentable]
}

extension RNSPGlobalCMPConsent {
    init?(from globalCMP: SPGlobalCmpConsent?) {
        guard let globalCMP = globalCMP else { return nil }

        self.init(
          uuid: globalCMP.uuid,
          applies: globalCMP.applies,
          vendors: globalCMP.vendors.map { RNSPConsentable(from: $0) },
          categories: globalCMP.categories.map { RNSPConsentable(from: $0) }
        )
    }
} 
